File: C++/encode-string-with-shortest-length.cpp
// Time:  O(n^3) on average
// Space: O(n^2)

class Solution {
public:
    string encode(string s) {
        vector<vector<string>> dp(s.length(), vector<string>(s.length()));
        for (int len = 1; len <= s.length(); ++len) {
            for (int i = 0; i + len - 1 < s.length(); ++i) {
                int j = i + len - 1;
                dp[i][j] = s.substr(i, len);
                for (int k = i; k < j; ++k) {
                    if (dp[i][k].length() + dp[k + 1][j].length() < dp[i][j].length()) {
                        dp[i][j] = dp[i][k] + dp[k + 1][j];
                    }
                }
                string encoded_string = encode_substr(dp, s, i, j);
                if (encoded_string.length() < dp[i][j].length()) {
                    dp[i][j] = encoded_string;
                }
            }
        }
        return dp[0][s.length() - 1];
    }

private:
    string encode_substr(const vector<vector<string>>& dp, const string& s, int i, int j) {
        string temp = s.substr(i, j - i + 1);
        auto pos = (temp + temp).find(temp, 1);  // O(n) on average
        if (pos >= temp.length()) {
            return temp;
        }
        return to_string(temp.length() / pos) + '[' + dp[i][i + pos - 1] + ']';
    }
};

File: Python/encode-string-with-shortest-length.py
# Time:  O(n^3) on average
# Space: O(n^2)

class Solution(object):
    def encode(self, s):
        """
        :type s: str
        :rtype: str
        """
        def encode_substr(dp, s, i, j):
            temp = s[i:j+1]
            pos = (temp + temp).find(temp, 1)  # O(n) on average
            if pos >= len(temp):
                return temp
            return str(len(temp)/pos) + '[' + dp[i][i + pos - 1] + ']'

        dp = [["" for _ in xrange(len(s))] for _ in xrange(len(s))]
        for length in xrange(1, len(s)+1):
            for i in xrange(len(s)+1-length):
                j = i+length-1
                dp[i][j] = s[i:i+length]
                for k in xrange(i, j):
                    if len(dp[i][k]) + len(dp[k+1][j]) < len(dp[i][j]):
                        dp[i][j] = dp[i][k] + dp[k+1][j]
                encoded_string = encode_substr(dp, s, i, j)
                if len(encoded_string) < len(dp[i][j]):
                    dp[i][j] = encoded_string
        return dp[0][len(s) - 1]

